src/burnchains/bitcoin/indexer.rs

```diff
@@ -104,155 +117,240 @@ pub struct BitcoinIndexer {
+            config: self.config.clone(),
+            runtime: BitcoinIndexerRuntime::new(self.runtime.network_id)
+        }
     }
 
     /// (re)connect to our configured network peer.
     /// Sets self.runtime.sock to a new socket referring to our configured
     /// Bitcoin peer.  If we fail to connect, this method sets the socket
     /// to None.
-    fn reconnect_peer(&mut self) -> Result<(), &'static str> {
+    fn reconnect_peer(&mut self) -> Result<(), btc_error> {
         match net::TcpStream::connect((self.config.peer_host.as_str(), self.config.peer_port)) {
             Ok(s) => {
                 self.runtime.sock = Arc::new(Mutex::new(Some(s)));
-                return Ok(());
+                Ok(())
             },
             Err(_e) => {
                 self.runtime.sock = Arc::new(Mutex::new(None));
-                return Err("Failed to connect to remote peer");
+                Err(btc_error::ConnectionError)
             }
         }
     }
 
     /// Get a locked handle to the internal socket 
     pub fn socket_locked(&mut self) -> LockResult<MutexGuard<Option<net::TcpStream>>> {
-        return self.runtime.sock.lock();
+        self.runtime.sock.lock()
     }
 
     /// Open an RPC connection to bitcoind 
     pub fn get_bitcoin_client(&self) -> BitcoinRPC {
-        let client = BitcoinRPC::new(
-            format!("http://{}:{}", self.config.peer_host.as_str(), self.config.rpc_port),
+        BitcoinRPC::new(
+            format!("{}://{}:{}", if self.config.rpc_ssl { "https" } else { "http" }, self.config.peer_host.as_str(), self.config.rpc_port),
             self.config.username.clone(),
             self.config.password.clone()
-        );
-        return client;
+        )
+    }
+
+    /// Are we connected?
+    fn is_connected(&mut self) -> bool {
+        let sock_locked = self.socket_locked();
+        match sock_locked {
+            Err(_) => {
+                false
+            },
+            Ok(mut guard) => {
+                match *guard.deref_mut() {
+                    Some(ref mut sock) => {
+                        true
+                    },
+                    None => {
+                        false
+                    }
+                }
+            }
+        }
     }
 
     /// Carry on a conversation with the bitcoin peer.
     /// Handle version, verack, ping, and pong messages automatically.
     /// Reconnect to the peer automatically if the peer closes the connection.
     /// Pass any other messages to a given message handler.
-    pub fn peer_communicate<T: BitcoinMessageHandler>(&mut self, message_handler: &mut T) -> Result<(), btc_error> {
-        let mut do_handshake = true;
+    pub fn peer_communicate<T: BitcoinMessageHandler>(&mut self, message_handler: &mut T, initial_handshake: bool) -> Result<(), btc_error> {
+        let mut do_handshake = initial_handshake || !self.is_connected();
         let mut keep_going = true;
         let mut initiated = false;
 
```

Comment:
if there's an error in `self.connect_handshake_backoff`, it should probably be logged, right? since that function retries in ConnectionBroken cases automatically, if it returns an error, it likely means that there's something pretty wrong.

Reply:
It is logged in `connect_handshake_backoff`.  Is that what you mean?

The intent is indeed to loop forever -- the node is already doing this by continuously trying to find later and later Bitcoin blocks, and it should not stop doing so even if the network goes down (since the network will likely come back up later).  The design question is what we want the main loop to look like?

Option 1:

```
while True:
   new_height = get_blockchain_height();
   rc = sync_blockchain(new_height);
   if not rc:
      break

   sleep(delay)
```

Option 2:

```
while true:
   new_height = get_blockchain_height();
   had_failure = false;
   while still_running():
      rc = sync_blockchain(new_height);
      if is_network_error(rc):
         # try again 
         had_failure = true;
         continue
      else:
         break
   if had_network_failure:
      # don't sleep -- the burn chain may have advanced
      continue

   sleep(delay)
```

Option 1 is what's implemented now.  The top-level call `sync_blockchain` internally masks network failures and retries fetching blocks until it succeeds in reaching the `new_height`.

Option 2 is what we would need to do if `connect_handshake_backoff` did not loop forever -- i.e. `sync_blockchain` would error immediately in the event of connectivity problems.

I'm happy with either approach -- no state-changes are effected either way unless the system successfully fetches actionable data from the peer network.